### PR TITLE
HPCC-15682 Add auto project for unprojected indexes

### PR DESCRIPTION
**Pull request by ghalliday**
Signed-off-by: Gavin Halliday gavin.halliday@lexisnexis.com

### CODE REVIEW

**Comment by ghalliday** [comment]
@jamienoss please review

File: ecl/hql/hqlpmap.hpp [hunk 1]
```diff
@@ -75,6 +74,7 @@ class HQL_API NewProjectMapper2 : public CInterface
 
     IHqlExpression * recursiveExpandSelect(IHqlExpression * expr, IHqlExpression * oldDataset, IHqlExpression * newDataset, IHqlExpression * oldParent);
     IHqlExpression * doNewExpandSelect(IHqlExpression * expr, IHqlExpression * oldDataset, IHqlExpression * newDataset, IHqlExpression * oldParent);
+    IHqlExpression * expandKeyedSelectors(IHqlExpression * expr, IHqlExpression * oldDataset, IHqlExpression * newDataset);
```
Comment (jamienoss) [inline ecl/hql/hqlpmap.hpp]: Why was this added?

File: ecl/hql/hqlpmap.cpp [hunk 2]
```diff
@@ -524,8 +524,15 @@ IHqlExpression * NewProjectMapper2::recursiveExpandSelect(IHqlExpression * expr,
         unsigned match = targets.find(*plain);
         if (match == NotFound)
         {
-            if (ignoreMissingFields)
-                return LINK(expr);
+            /*
+             * Sometimes the implicit project code will remove fields that are used in WILD() expressions.  This creates an inconsistency
+             * in the graph - a reference to a field that doesn't exist in the output of the previous expression.  If we are expanding
+             * a selector reference, and it isn't found in the input's dataset, assume it has been projected out, and remap it to the
+             * parent dataset.  (If this is wrong it will trigger an internal error later on.)
```
Comment (jamienoss) [inline ecl/hql/hqlpmap.cpp]: Are we sure of this assertion - "(If this is wrong it will trigger an internal error later on.)" for all scenarios?

File: ecl/hql/hqlpmap.hpp [hunk 3]
```diff
@@ -32,7 +32,7 @@ class TableProjectMapper;
 class HQL_API NewProjectMapper2 : public CInterface
 {
 public:
-    NewProjectMapper2() { ignoreMissingFields = false; mapping = NULL; matchedAll = true; expandCallback = NULL; }
+    NewProjectMapper2() { mapping = NULL; matchedAll = true; expandCallback = NULL; }
```
Comment (jamienoss) [inline ecl/hql/hqlpmap.hpp]: NULL -> nullptr

**Comment by jamienoss** [comment]
@ghalliday - a few comments, back to you.

File: ecl/hql/hqlpmap.hpp (hunk 1 above)
Comment (ghalliday) [inline ecl/hql/hqlpmap.hpp]: I'll remove - that was from a first pass.

File: ecl/hql/hqlpmap.cpp (hunk 2 above)
Comment (ghalliday) [inline ecl/hql/hqlpmap.cpp]: I am sure enough to include the code and the comment.  There might be situations that aren't caught, but I can't conceive of any.  Do you have an alternative suggestion?

File: ecl/hql/hqlpmap.hpp (hunk 3 above)
Comment (ghalliday) [inline ecl/hql/hqlpmap.hpp]: not new code, so I didn't change.

**Comment by ghalliday** [comment]
@jamienoss line deleted.  See other replies to other comments.

File: ecl/hql/hqlpmap.cpp (hunk 2 above)
Comment (jamienoss) [inline ecl/hql/hqlpmap.cpp]: I have none other than to look further. One of my concerns is whether it is possible that future code changes will require knowledge of this PR to prevent creating a scenario that isn't caught because it relies on other assertions rather than this code change having it's own. This is ok if the other assertions are core/base/foundation assertions but if they are not... It's a question of containment.

**Comment by jamienoss** [comment]
@richardkchapman other than my comment regarding containment this looks ok to me.
